063-topic-replication-factor-change.md

```diff
@@ -0,0 +1,240 @@
+Currently, the topic configuration operation only supports changing the replication factor.
+Under the cover, Cruise Control reuse the same heuristic used for the rebalance operation to determine the best location to place or remove partition replicas.
+The Cluster Operator is already integrated with Cruise Control, so we can factor out the existing logic and classes.
+
+### Cruise Control
+
+Cruise Control creates a task for any operation, such as topic configurations and cluster rebalances.
+A task is first created in Active state, then moves to InExecution, and finally to Completed or CompletedWithError.
+At most `max.active.user.tasks` tasks can be Active at the same time (5 by default), and only one task can be InExecution at any given time.
+
+#### Topic configuration endpoint
+
+The `topic_configuration` endpoint is used to request a replication factor change.
+Once accepted, the replication factor change is put into the Cruise Control's task queue and executed asynchronously.
+
+The two required parameters are `topic`, which is a regular expression to specify subject topics, and `replication_factor`, which is the target replication factor.
+In order to batch multiple changes in the same request, the Topic Operator uses a JSON payload containing topics grouped by their replication factor.
+
+```sh
+$ curl -vv -X POST -H "Content-Type: application/json" -d '{replication_factor:{topic_by_replication_factor:{2:topic1|topic3|topic4,3:topic2|topic5}}}' \
+  "http://localhost:9090/kafkacruisecontrol/topic_configuration?skip_rack_awareness_check=true&dryrun=false&json=true" | jq
+> POST /kafkacruisecontrol/topic_configuration?skip_rack_awareness_check=true&dryrun=false&json=true HTTP/1.1
+> Host: localhost:9090
+> User-Agent: curl/8.0.1
+> Accept: */*
+> 
+} [91 bytes data]
+< HTTP/1.1 200 OK
+< Date: Thu, 07 Dec 2023 14:58:55 GMT
+< Set-Cookie: JSESSIONID=node0lfkx183ijq4pjkkjahsegfmp30.node0; Path=/
+< Expires: Thu, 01 Jan 1970 00:00:00 GMT
+< User-Task-ID: 5344ca89-351f-4565-8d0f-9aade00e053d
+< Content-Type: application/json;charset=utf-8
+< Cruise-Control-Version: 2.5.77-SNAPSHOT
+< Cruise-Control-Commit_Id: 535ea07e8847990cc02dde1f1be99b387dbeaf5b
+< Content-Length: 17051
+< Server: Jetty(9.4.52.v20230823)
+< 
+{ [17321 bytes data]
+{
+  "summary": {
+    "numIntraBrokerReplicaMovements": 0,
+    "numReplicaMovements": 125,
+    "onDemandBalancednessScoreAfter": 100,
+    "intraBrokerDataToMoveMB": 0,
+    "monitoredPartitionsPercentage": 100,
+    "provisionRecommendation": "",
+    "excludedBrokersForReplicaMove": [],
+    "excludedBrokersForLeadership": [],
+    "provisionStatus": "RIGHT_SIZED",
+    "onDemandBalancednessScoreBefore": 100,
+    "recentWindows": 5,
+    "dataToMoveMB": 0,
+    "excludedTopics": [],
+    "numLeaderMovements": 0
+  }
+  ...
+```
+
+The `goals` parameter is the comma separated list of goals used to generate the configuration proposal.
```

Comment:
You mention the rebalance proposal here. But do not cover it later. Will it be used by TO? Or will TO directly change the RF? It would be worth explaining.

Reply:
TO requests the RF change to CC. Quoted from the proposal: "Under the cover, Cruise Control reuse the same heuristic used for the rebalance operation to determine the best location to place or remove partition replicas."

Reply:
Sorry, maybe my comment here wasn't clear. What the sentence in the proposal suggests to me is that CC gives you a proposal of how it will add the replicas and you approve it. Is that the case or not?

Reply:
No, it's all automatic. I'll add a clarification.